### ISSUE TITLE
Mouse wheel scroll inserts text into the command window

### ISSUE THREAD

**Issue by frasercrmck** [neovim/neovim#2936]
Every so often when scrolling a buffer using the mouse wheel I get text inserted into the command window for me, which stops me scrolling as focus has shifted. The text looks like this: `.,.+6455;33M`. The numbers aren't always the same, but the rest is.

This only started after updating neovim to the current head, b656a95, from one a few days before that. 

I'm on ArchLinux and I also had to change a few AUR packages during the update: It removed `libvterm-git-neovim` and installed `libvterm-bzr` in its place.

**Comment by lucc** [comment]
I also noticed this. I am also on arch linux and did the resent upgrade of neovim-git and switched from libvterm-git-neovim to libvterm-bzr. I am currently on b4d45f6 and the text which is inserted for me is different. Will post details later.

**Comment by pallavagarwal07** [comment]
+1
I'm also experiencing the same proble. I'm also on Arch linux, on Urxvt terminal. It also happens irrespective of whether I'm using tmux or not.

**Comment by fwalch** [comment]
Just a wild guess: does it go away if you `:set mouse=`?

**Comment by pallavagarwal07** [comment]
The text doesn't go away.  But obviously since the scroll stops working, the bug isn't encountered anymore. 
In vim since mouse isn't on by default, I checked with `set mouse=a`

**Comment by fwalch** [comment]
So you had `set mouse=a` in your `.nvimrc`? Or did you just started using it now, when it became the default (#2860)? I'm trying to understand when/how this has been introduced.

**Comment by pallavagarwal07** [comment]
I added `set mouse=nv` to vimrc when mouse became default, and I hated mouse in insert mode. But I've tested in both vim and nvim with both a and nv options. I've been using nvim for about two weeks and I think this problem only started with a recent update.

**Comment by lucc** [comment]
I had `mouse=a` in my nvimrc until it became the default.

**Comment by lucc** [comment]
I had `mouse=a` in my nvimrc until it became the default.

**Comment by fwalch** [comment]
If this problem only started only after the most recent Neovim update, can you check your pacman log (`/var/log/pacman.log`) to see which version you used previously? There should be a line like

    upgraded neovim-git (0.r172.e18c075-1 -> 0.r214.f479abd-1)

**Comment by pallavagarwal07** [comment]
I think it's this one. 
0.r3570.0685489-1  - > 0.r3619.4aa6279-1

**Comment by lucc** [comment]
I just used `git bisect` to find the commit that introduces this, like this:
```sh
git bisect 0685489^ master # git bisect does not like 0685 so we take the parent
git bisect run sh -c "make && build/bin/nvim -u NONE -i NONE -c 'set mouse=a' src/nvim/eval.c"
# now scroll around a lot and use :

**Comment by lucc** [comment]
I just used `git bisect` to find the commit that introduces this, like this:
```sh
git bisect master 0685489 master
git bisect run sh -c "make && build/bin/nvim -u NONE -i NONE -c 'set mouse=a' src/nvim/eval.c"
# now scroll around a lot and use :q or :cq to quit with exit status 0 or 1
```
That resulted in dcaf9c6bc3d5f83782fca7a145ba5feac7746b1e being the last good and 0ef80b9c2b922280c3ba2c0a8638f23ae57d6618 being the first bad revision (both where merge into master together at b656a954cfec0379a5bb2f7f5d1f28cbc03324df).

**Comment by fwalch** [comment]
Since 0ef80b9 was part of #2650, ping @tarruda.

**Comment by oni-link** [comment]
If the mouse wheel is scrolled fast, two inputs are generated. When searching for the end of the first input in the input buffer, the end is off by one. The following patch fixes the described problem for me.
```diff
diff --git a/src/nvim/tui/term_input.inl b/src/nvim/tui/term_input.inl
index 451ac19..9c62362 100644
--- a/src/nvim/tui/term_input.inl
+++ b/src/nvim/tui/term_input.inl
@@ -243,6 +243,7 @@ static void read_cb(RStream *rstream, RBuffer *buf, void *data, bool eof)
     RBUFFER_EACH(input->read_buffer, c, i) {
       count = i + 1;
       if (c == '\x1b' && count > 1) {
+        count--;
         break;
       }
     }
```

**Comment by lucc** [comment]
The patch indeed fixes it for me and all tests still pass.
